core/src/sim.c

```diff
@@ -290,36 +277,116 @@ int handle_potential_stopping_condition(Simmer *simmer) {
+  Rack *rack_placeholder = simmer_worker->rack_placeholder;
+  Simmer *simmer = simmer_worker->simmer;
+  int plies = simmer->max_plies;
+
+  // set random rack for opponent (throw in rack, shuffle, draw new tiles).
+  set_random_rack(game, 1 - game->player_on_turn_index, simmer->known_opp_rack);
+  // need a new shuffle for every iteration:
+  shuffle(game->gen->bag);
+
+  for (int i = 0; i < simmer->num_simmed_plays; i++) {
+    if (simmer->simmed_plays[i]->ignore) {
+      continue;
+    }
+
+    double leftover = 0.0;
+    set_backup_mode(game, BACKUP_MODE_SIMULATION);
+    // play move
+    play_move(game, simmer->simmed_plays[i]->move);
+    simmer_worker->node_count++;
+    set_backup_mode(game, BACKUP_MODE_OFF);
+    // further plies will NOT be backed up.
+    for (int ply = 0; ply < plies; ply++) {
+      int onturn = game->player_on_turn_index;
+      if (game->game_end_reason != GAME_END_REASON_NONE) {
+        // game is over.
+        break;
       }
-      if (tc->simmer->ucgi_mode == UCGI_MODE_ON &&
-          (total_iterations % 100) < tc->simmer->threads) {
-        // Print out an estimate of how it's going in UCGI format.
-        print_ucgi_stats(tc->simmer, 0);
+
+      Move *best_play = get_top_equity_move(game);
+      copy_rack_into(rack_placeholder, game->players[onturn]->rack);
+      play_move(game, best_play);
+      simmer_worker->node_count++;
+      char placeholder[80];
+      store_move_description(best_play, placeholder,
+                             game->gen->letter_distribution);
+
+      if (ply == plies - 2 || ply == plies - 1) {
+        double this_leftover =
+            get_leave_value_for_move(game->players[0]->strategy_params->klv,
+                                     best_play, rack_placeholder);
+        if (onturn == simmer->initial_player) {
+          leftover += this_leftover;
+        } else {
+          leftover -= this_leftover;
+        }
       }
+      add_score_stat(simmer->simmed_plays[i], best_play->score,
+                     best_play->tiles_played == 7, ply,
+                     is_multithreaded(simmer));
+    }
+
+    int spread = game->players[simmer->initial_player]->score -
+                 game->players[1 - simmer->initial_player]->score;
+    add_equity_stat(simmer->simmed_plays[i], simmer->initial_spread, spread,
+                    leftover, is_multithreaded(simmer));
+    add_winpct_stat(
+        simmer->simmed_plays[i], simmer->win_pcts, spread, leftover,
+        game->game_end_reason,
+        // number of tiles unseen to us: bag tiles + tiles on opp rack.
+        game->gen->bag->last_tile_index + 1 +
+            game->players[1 - simmer->initial_player]->rack->number_of_letters,
+        plies % 2, is_multithreaded(simmer));
+    // reset to first state. we only need to restore one backup.
+    unplay_last_move(game);
+  }
+}
+
+void *simmer_worker(void *uncasted_simmer_worker) {
+  SimmerWorker *simmer_worker = (SimmerWorker *)uncasted_simmer_worker;
```

Comment:
this cast is not technically necessary I learned. According to my copy of K&R it doesn't seem to be needed 😝

Reply:
The compiler gives warnings if this isn't included.